File: src/bytecode/Deserializer.ts
import { MemoryStream } from "../util/MemoryStream";
import { Chunk } from "./structs/Chunk";

export class Deserializer {
  byteStream: MemoryStream;

  constructor(bytes: number[]) {
    this.byteStream = new MemoryStream(bytes);
  }

  deserialize(): Chunk {
    if (this.byteStream.readString(4) != String.fromCharCode(27) + "Lua")
      throw new Error("Invalid file header");

    if (this.byteStream.readInt8() != 0x51)
      throw new Error("Invalid lua version, only Lua 5.1 supported");

    if (this.byteStream.readInt8() != 0)
      throw new Error("Invalid format versions");

    //TODO: Big Endian Support
    //Endianess value (1 = little 0 = big)
    if (this.byteStream.readInt8() != 1) throw new Error("Invalid endianess");

    this.byteStream.intSize = this.byteStream.readInt8();
    if (this.byteStream.intSize != 4 && this.byteStream.intSize != 8)
      throw new Error("Invalid int size");

    this.byteStream.sizeTSize = this.byteStream.readInt8();
    if (this.byteStream.sizeTSize != 4 && this.byteStream.sizeTSize != 8)
      throw new Error("Invalid SizeT size");

    if (this.byteStream.readInt8() != 4)
      throw new Error("Invalid Instruction Size");

    //TODO: Add float (4-bit) number support
    if (this.byteStream.readInt8() != 8)
      throw new Error("Invalid lua number size");

    if (this.byteStream.readInt8() != 0)
      throw new Error("Invalid integral flag");

    return new Chunk(this.byteStream);
  }
}

File: lua_deserializer/src/deserializer.rs
use crate::{structs::chunk::Chunk, util::read_stream::ReadStream};

pub struct Deserializer {
    memory_stream: ReadStream,
}

impl Deserializer {
    pub fn new(bytes: Vec<u8>) -> Self {
        Self {
            memory_stream: ReadStream::new(bytes),
        }
    }

    pub fn deserialize(&mut self) -> Chunk {
        assert_eq!(
            self.memory_stream.read_string_length(4),
            String::from_utf8(vec![27]).unwrap() + "Lua",
            "Invalid file header"
        );

        assert_eq!(
            self.memory_stream.read_int8(),
            0x51,
            "Invalid Lua version, only Lua 5.1 supported"
        );

        assert_eq!(self.memory_stream.read_int8(), 0, "Invalid format version");

        //TODO: Alternate endianess
        // Endianess value (1 = little 0 = big)
        assert_eq!(self.memory_stream.read_int8(), 1, "Invalid endianess value");

        self.memory_stream.int_size = self.memory_stream.read_int8();
        assert!(
            self.memory_stream.int_size == 4 || self.memory_stream.int_size == 8,
            "Invalid int size"
        );

        self.memory_stream.size_t_size = self.memory_stream.read_int8();
        assert!(
            self.memory_stream.size_t_size == 4 || self.memory_stream.size_t_size == 8,
            "Invalid size t size"
        );

        assert_eq!(
            self.memory_stream.read_int8(),
            4,
            "Invalid instruction size"
        );

        //TODO: Float (4-bit) number support
        assert_eq!(self.memory_stream.read_int8(), 8, "Invalid lua number size");

        assert_eq!(self.memory_stream.read_int8(), 0, "Invalid integral flag");

        Chunk::new(&mut self.memory_stream)
    }
}
